codegen/src/io/enums.rs

```diff
@@ -0,0 +1,361 @@
+        // ie:
+        // enum MyEnum {
+        //   Test(HERE)
+        // }
+        variants.push(parse_enum_variant(variant, &attributes, &discrim, error_stream));
+
+        // this is hacky, but we need to check whether or not the error_stream has any errors.
+        // if it does, we need to return an empty token stream.
+        if !error_stream.is_empty() {
+            return TokenStream::new();
+        }
+    }
+
+    // get all write streams from variants
+    let write_streams = variants.iter().map(|variant| variant.write_content.clone()).collect::<Vec<TokenStream2>>();
+    let read_streams = variants.iter().map(|variant| variant.read_content.clone()).collect::<Vec<TokenStream2>>();
+
+    quote! {
+        impl ::binary_utils::interfaces::Writer for #enum_name {
+            fn write(&self, _binary_writew: &mut ::binary_utils::io::ByteWriter) -> ::std::result::Result<(), ::std::io::Error> {
+                match self {
+                    #(#write_streams)*
+                };
+
+                Ok(())
+            }
+        }
+
+        impl ::binary_utils::interfaces::Reader<#enum_name> for #enum_name {
+            fn read(_binary_readerr: &mut ::binary_utils::io::ByteReader) -> ::std::result::Result<#enum_name, ::std::io::Error> {
+                match <#repr_type>::read(_binary_readerr)? {
+                    #(#read_streams)*
+                    _ => Err(::std::io::Error::new(::std::io::ErrorKind::InvalidData, "Invalid enum discriminant."))
+                }
+            }
+        }
+    }.into()
+}
+
+fn parse_enum_variant(
+    variant: &syn::Variant,
+    _attributes: &Vec<super::util::attrs::IoAttr>,
+    curr_discrim: &syn::LitInt,
+    error_stream: &mut TokenStream2,
+) -> ParsedEnumVariant {
+    let mut read_content = TokenStream2::new();
+    let mut write_content = TokenStream2::new();
+
+    let variant_name = &variant.ident;
+
+    match variant.fields {
+        Fields::Unnamed(ref fields) => {
+            // This is the stream within the match arm
+            let mut read_inner = TokenStream2::new();
+            let mut write_inner = TokenStream2::new();
+
+            let mut args: Vec<syn::Ident> = Vec::new();
+
+            for (i, field) in fields.unnamed.iter().enumerate() {
+                let inner_attrs = field
+                    .attrs
+                    .iter()
+                    .filter_map(|att| {
+                        match parse_attribute(&att, error_stream) {
+                            Ok(attr) => Some(attr),
+                            Err(_) => None,
+                        }
+                    })
+                    .collect::<Vec<super::util::attrs::IoAttr>>();
+
+                if inner_attrs.len() != 0 {
+                    error_stream.append_all(
+                        syn::Error::new_spanned(
+                            field,
+                            "Attributes are not valid on enum variant fields at this time.",
```

Comment:
I might open an issue for this, however im not sure what derive attributes would be useful on enums at this time.